### ISSUE TITLE
several issues with "color" tab in background chooser

### ISSUE THREAD

**Issue by jeisner** [mypaint/mypaint#469]
There are a few things that don't work quite as I'd expect.  Here's my proposal:

* If the current background is a solid color, then opening or switching to the Color tab should set the selected color to the current background color.  
  * Reason: **This makes it easier to fine-tune the current color.** 
  * Note: A solid-color background may be in effect even if the Color tab hasn't been used.  It might have been set by a file load or by a solid-color choice on the Pattern tab.  Solid colors set in these ways should be fine-tunable too -- the user shouldn't have to remember where the color came from.

* Even if the current background is not a solid color, opening or switching to the Color tab should set the selected color to the most recent solid background color.  
  * The previous case can be regarded as a special case of this.  
  * Again, solid backgrounds still count as solid even if they came from a file load or the pattern tab.
  * I don't think "most recent solid background color" needs to be saved or loaded on a per-drawing basis.  The simplest approach is for MyPaint to just remember what that color was, and save it across sessions in the config file.  On the very first run of MyPaint, the background defaults to solid white, so that will presumably be the initial value.  
  * Note: At present, every run of MyPaint seems to reinitialize the Color tab to black for some reason.

* Opening or switching to the Color tab should not only select a color, but should also set the actual background to match this selected color.  
  * Note: Ordinarily the actual background tracks the selected color, **so it seems weird that at present, they don't match when you enter the tab in the first place.**  Indeed, it's actually a bit hard to force the background to switch to the selected color at first - you have to either use the eyedropper or edit the color slightly.

* There are two color swatches in the lower left of the tab.  I'd suggest that the *left* swatch should be the current color, and the *right* swatch should be the color that was selected the last time that we opened or switched to the Color tab.  Furthermore, it should be possible to click on that right swatch (without using the eyedropper) in order to restore that originally selected color.  **This design is meant to resemble the two color swatches in the lower left of the paint window.**
  * Note: At present, it seems that the left swatch is always black (why is that useful?), and the right swatch is the current color.

**Comment by 0ion9** [comment]
First point :+1: 

Second point : (gamma-correctly) averaging the pixels of the background tile seems like it would be generally more intuitive than jumping back to a previously stored color. On the subject of previously stored colors, a little 'palette' showing the last N (where N is, let's say, 16) selected background colors (much like the color history in GIMP's color editor), that can be clicked on to set that as the BG, would be clear without being 'automagical' (which automatically jumping back to a stored value feels like, to me)

Third and fourth points: :+1:  I think.

.. Left swatch hasn't always been black, I'm pretty sure it did show the previous color. Unfortunately mypaint doesn't currently run for me so I can't check the current state.

**Comment by jeisner** [comment]
> without being 'automagical' (which automatically jumping back to a stored value feels like, to me)

Hmm, it didn't seem strange to me.  I just thought of it as opening up the Color tab and finding it in the state where you last left it.  

Choosing a solid color from *outside* the color tab (by choosing a solid pattern or loading a file) is like sending a friend into the Color tab to change it for you.  So when you go back to it, you find it in the state where your friend last left it.

> (gamma-correctly) averaging the pixels of the background tile seems like it would be generally more intuitive than jumping back to a previously stored color.

I'd thought about this too, but I'm not sure that averaging really does what you want.  Some of the backgrounds are textured paper with light and shadow; the perceived color of the paper usually comes from the light parts and shouldn't be averaged with the dark parts.  Other backgrounds are lined paper; there the perceived color is the paper color, not an average with the color of the lines.

You could attempt to make a good guess using a heuristic that is fancier than averaging.  But fortunately, the eyedropper in the background Color tab already allows the user to match any *chosen* pixel of the current background.  That puts the user in control, and I thought that might be enough for non-solid backgrounds.  

Note that switching from a pattern to a similar solid color is a *discrete* change and perhaps won't happen that often; the eyedropper may be the best way to manage it when needed.  By contrast, fine-tuning a current solid color in *continuous* color space seems like a common operation and one that we should support.   After all, after the user chooses a solid color, it's easy to get slightly dissatisfied with how it looks later, once there's a drawing on top of it.   

Currently there is a sharp difference between non-solid and solid backgrounds: *only* the solid backgrounds can be continuously adjusted, and the Color tab is used *only* for that.   This is why I think the Color tab should remember the last solid background.  But it would be a cool and useful future enhancement if the Color tab could be used to recolor any background (I've created this as #471).  Under that design, I would agree that the Color tab should *not* switch back to the previous solid color (because there would no longer be anything special about solid colors).  Rather, the Color tab would show a color that was derived from the current background pattern ... and adjusting that color would indirectly adjust *all* colors in the pattern in some consistent way (in effect, it would specify a transformation of HSV space).  

> On the subject of previously stored colors, a little 'palette' showing the last N ...

Yeah.  I've posted this for you as a separate issue, #472.

> .. Left swatch hasn't always been black, I'm pretty sure it did show the previous color. Unfortunately mypaint doesn't currently run for me so I can't check the current state.

Perhaps a bug got introduced?  At any rate, shouldn't the previous color have been the *right* swatch, to match the design for the swatches at the lower left of the *main* paint window?

(Here's the current design for the main paint window: The right swatch is the previous stroke's color.  The left swatch is your current choice for the next stroke's color, and can be updated with color choosers.  When you start a new stroke, it's in the color of the left swatch; when you finish that stroke, the right swatch is updated to match the left swatch.  By clicking on the right swatch, you can make the left swatch match the right swatch.)

**Comment by odysseywestra** [comment]
@briend @jplloyd This probably an issue that should be addressed before a 2.0. This does break UX usability.

**Comment by jplloyd** [comment]
Addressing points:
**1.** Agreed, but we would either have to amend the file format, include this in the bundled settings, or use the solution similar to the one proposed by @0ion9.

**2.** Possibly, but I think having a selection of favorites would be the better solution.

**3.** Disagree, this would make it impossible/tedious to select a color from a previous patterned background. If **1.** and the second half of **4.** is implemented, changing the color with single a click is fast enough.

**4.** The left swatch indeed seems to be useless. This should be fixed.

On the whole I would contest that this *breaks* UX usability, even though it can be a bit inconvenient.
At any rate, if @odysseywestra or @briend  (or anyone else for that matter) wants to take this on for 2.0, that's fine. If not, it's deferred to 2.1.0 at the earliest.

**Comment by odysseywestra** [comment]
Yeah, if this is going to have to require to amend the .ora file spec, I would deter this till 2.1.0 then.
